model/system_metrics_test.go

```diff
@@ -70,6 +77,166 @@ func TestSystemMetricsFind(t *testing.T) {
 	})
 }
 
+func TestSystemMetricsAppend(t *testing.T) {
+	env := cedar.GetEnvironment()
+	db := env.GetDB()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	tmpDir, err := ioutil.TempDir(".", "append-test")
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, os.RemoveAll(tmpDir))
+		assert.NoError(t, db.Collection(systemMetricsCollection).Drop(ctx))
+		assert.NoError(t, db.Collection(configurationCollection).Drop(ctx))
+	}()
+
+	testBucket, err := pail.NewLocalBucket(pail.LocalOptions{Path: tmpDir})
+	require.NoError(t, err)
+
+	systemMetrics := getSystemMetrics()
+	systemMetrics.populated = true
+
+	chunk1 := make([]byte, 32)
+	chunk2 := make([]byte, 32)
+	rand.Read(chunk1)
+	rand.Read(chunk2)
+
+	t.Run("NoEnv", func(t *testing.T) {
+		sm := SystemMetrics{ID: systemMetrics.ID}
+		assert.Error(t, sm.Append(ctx, 1, chunk1))
+	})
+	t.Run("DNE", func(t *testing.T) {
+		systemMetrics.Setup(env)
+		assert.Error(t, systemMetrics.Append(ctx, 1, chunk1))
+	})
+	_, err = db.Collection(systemMetricsCollection).InsertOne(ctx, systemMetrics)
+	require.NoError(t, err)
+	t.Run("NoConfig", func(t *testing.T) {
+		systemMetrics.Setup(env)
+		assert.Error(t, systemMetrics.Append(ctx, 1, chunk1))
+	})
+	conf := &CedarConfig{populated: true}
+	conf.Setup(env)
+	require.NoError(t, conf.Save())
+	t.Run("ConfigWithoutBucket", func(t *testing.T) {
+		systemMetrics.Setup(env)
+		assert.Error(t, systemMetrics.Append(ctx, 1, chunk1))
+	})
+	conf.Setup(env)
+	require.NoError(t, conf.Find())
+	conf.Bucket.SystemMetricsBucket = tmpDir
+	require.NoError(t, conf.Save())
+	t.Run("AppendToBucketAndDB", func(t *testing.T) {
+		systemMetrics.Setup(env)
+		require.NoError(t, systemMetrics.Append(ctx, 1, chunk1))
+		require.NoError(t, systemMetrics.Append(ctx, 2, chunk2))
+		expectedData := map[string][]byte{
+			"1": chunk1,
+			"2": chunk2,
+		}
+
+		b := &backoff.Backoff{
+			Min:    100 * time.Millisecond,
+			Max:    5 * time.Second,
+			Factor: 2,
+		}
+		var actualData map[string][]byte
+		for i := 0; i < 10; i++ {
+			actualData = map[string][]byte{}
+			iter, err := testBucket.List(ctx, systemMetrics.ID)
+			require.NoError(t, err)
+			for iter.Next(ctx) {
+				key, err := filepath.Rel(systemMetrics.ID, iter.Item().Name())
+				require.NoError(t, err)
+				r, err := iter.Item().Get(ctx)
+				require.NoError(t, err)
+				defer func() {
+					assert.NoError(t, r.Close())
+				}()
+				data, err := ioutil.ReadAll(r)
+				require.NoError(t, err)
+				actualData[key] = data
+			}
+
+			if len(actualData) > 1 {
+				break
+			}
+			time.Sleep(b.Duration())
```

Comment:
This is probably a @julianedwards question but why is the sleep + polling the bucket necessary here if everything is synchronous?

Reply:
There were so many flaky tests where on (mostly windows) it wouldn't write fast enough to the local bucket so Sam and I decided to add this terrible retry loop in buildlogger (Colin is copying that here).